File: mcp-server/llm.js
// Gemini API wrapper for server-side reasoning calls.
//
// Used by handlers that want LLM synthesis or "do you know this entity?"
// gating before paying SerpAPI quota. Lives separately from the
// extension's api.js because the contract is much narrower: structured
// JSON requests, low temperature, single-turn, no tool calls.
//
// Either GEMINI_API_KEY or GOOGLE_API_KEY (the Google AI Studio default
// name) is accepted. Read at call time so dotenv has had a chance to
// populate the env via index.js.

const GEMINI_MODEL   = "gemini-2.5-flash";
const GEMINI_API_URL = `https://generativelanguage.googleapis.com/v1beta/models/${GEMINI_MODEL}:generateContent`;

// Ask Gemini for a JSON-shaped response. The prompt should describe the
// expected schema explicitly. Returns the parsed object; throws if the
// response isn't valid JSON or the API call fails.
export async function geminiAskJson(prompt, opts = {}) {
  const key = process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY;
  if (!key) {
    throw new Error(
      "GEMINI_API_KEY (or GOOGLE_API_KEY) is not set in mcp-server/.env " +
      "(see .env.example). The server has no fallback for this path."
    );
  }
  if (!prompt || typeof prompt !== "string") {
    throw new Error("geminiAskJson requires a non-empty prompt string.");
  }

  const body = {
    contents: [{ role: "user", parts: [{ text: prompt }] }],
    generationConfig: {
      temperature: opts.temperature ?? 0,        // deterministic for structured tasks
      maxOutputTokens: opts.maxTokens ?? 1024,
      responseMimeType: "application/json"
    }
  };

  const res = await fetch(GEMINI_API_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-goog-api-key": key
    },
    body: JSON.stringify(body)
  });

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`Gemini ${res.status}: ${truncate(text || res.statusText, 250)}`);
  }

  const data = await res.json();
  const candidate = data.candidates?.[0];
  if (!candidate) {
    const reason = data.promptFeedback?.blockReason || "no candidates";
    throw new Error(`Gemini returned no usable response (${reason}).`);
  }

  const text = candidate.content?.parts?.map(p => p.text).filter(Boolean).join("") || "";
  if (!text) {
    throw new Error(`Gemini returned empty content (finishReason: ${candidate.finishReason || "unknown"}).`);
  }

  try {
    return JSON.parse(text);
  } catch {
    throw new Error(`Gemini did not return valid JSON: ${truncate(text, 200)}`);
  }
}

function truncate(s, n) {
  return s.length > n ? s.slice(0, n) + "…" : s;
}

File: mcp-server/llm.py
"""Gemini API wrapper for server-side reasoning calls.

Used by handlers that want LLM synthesis or "do you know this entity?"
gating before paying SerpAPI quota. Narrower contract than the
extension's api.js: structured JSON, low temperature, single turn,
no tool calls.

Either GEMINI_API_KEY or GOOGLE_API_KEY (the Google AI Studio default
name) is accepted. Read at call time so dotenv has had a chance to
populate the env at startup.
"""

from __future__ import annotations

import json
import os
from typing import Any

import httpx

GEMINI_MODEL = "gemini-2.5-flash"
GEMINI_API_URL = f"https://generativelanguage.googleapis.com/v1beta/models/{GEMINI_MODEL}:generateContent"

_HTTP_TIMEOUT = httpx.Timeout(60.0, connect=10.0)


async def gemini_ask_json(
    prompt: str, *, temperature: float = 0.0, max_tokens: int = 1024
) -> Any:
    """Ask Gemini for a JSON-shaped response. The prompt should describe the
    expected schema explicitly. Returns the parsed object; raises on
    invalid JSON or API failure."""
    key = os.environ.get("GEMINI_API_KEY") or os.environ.get("GOOGLE_API_KEY")
    if not key:
        raise RuntimeError(
            "GEMINI_API_KEY (or GOOGLE_API_KEY) is not set in mcp-server/.env "
            "(see .env.example). The server has no fallback for this path."
        )
    if not prompt or not isinstance(prompt, str):
        raise ValueError("gemini_ask_json requires a non-empty prompt string.")

    body: dict[str, Any] = {
        "contents": [{"role": "user", "parts": [{"text": prompt}]}],
        "generationConfig": {
            "temperature": temperature,
            "maxOutputTokens": max_tokens,
            "responseMimeType": "application/json",
        },
    }

    async with httpx.AsyncClient(timeout=_HTTP_TIMEOUT) as client:
        res = await client.post(
            GEMINI_API_URL,
            content=json.dumps(body),
            headers={
                "Content-Type": "application/json",
                "x-goog-api-key": key,
            },
        )

    if res.status_code != 200:
        snippet = _truncate(res.text or res.reason_phrase, 250)
        raise RuntimeError(f"Gemini {res.status_code}: {snippet}")

    data = res.json()
    candidates = data.get("candidates") or []
    if not candidates:
        reason = (data.get("promptFeedback") or {}).get("blockReason", "no candidates")
        raise RuntimeError(f"Gemini returned no usable response ({reason}).")

    candidate = candidates[0]
    parts = (candidate.get("content") or {}).get("parts") or []
    text = "".join(p.get("text") or "" for p in parts)
    if not text:
        finish = candidate.get("finishReason") or "unknown"
        raise RuntimeError(f"Gemini returned empty content (finishReason: {finish}).")

    try:
        return json.loads(text)
    except json.JSONDecodeError:
        raise RuntimeError(f"Gemini did not return valid JSON: {_truncate(text, 200)}")


def _truncate(s: str, n: int) -> str:
    return s if len(s) <= n else s[:n] + "…"
